{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from landscape_evolution import *\n",
    "import matplotlib as mpl\n",
    "import cProfile\n",
    "import pstats\n",
    "import time\n",
    "import random\n",
    "try:\n",
    "    import itertools.izip as zip\n",
    "except ImportError:\n",
    "    import itertools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from wFisher_func import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "simulate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is taken from Trevor Bedford's github. It is a Wright-Fisher simulation which includes: Mutation, Selection and Genetic Drift\n",
    "\n",
    "def mainfunc():\n",
    "\n",
    "    pop_size = 1000          # Size of Population\n",
    "    seq_length = 5            # This is \"N\"\n",
    "    generations = 10        # How long it runs\n",
    "    mutation_rate = 0.001     # per gen per individual per site\n",
    "    repeats = 10              # Number of landscape replicates\n",
    "\n",
    "    alphabet = ['0', '1']\n",
    "    base_haplotype = ''.join([\"0\" for i in range(seq_length)])\n",
    "\n",
    "    genotypes = [''.join(seq) for seq in itertools.product(\"01\", repeat=seq_length)]\n",
    "\n",
    "\n",
    "\n",
    "    pop = {}\n",
    "    fitness = {}\n",
    "    #fitnessB = {}\n",
    "\n",
    "    pop[base_haplotype] = pop_size\n",
    "    #pop2 = copy.deepcopy(pop)\n",
    "\n",
    "    A = Landscape(seq_length, 2)\n",
    "\n",
    "    A.ls = ((A.ls-np.min(A.ls))/(np.max(A.ls)-np.min(A.ls))) + 1e-5   # Normalize to fitness 0 --> 1 for fitness proportionality\n",
    "\n",
    "\n",
    "    for i in range(len(genotypes)):\n",
    "        fitness[genotypes[i]] = A.ls[i]\n",
    "        #fitnessB[genotypes[i]] = Bs[46].ls[i]\n",
    "\n",
    "    history = []\n",
    "    simulate(pop, history, generations, mutation_rate, pop_size, seq_length, fitness, alphabet)\n",
    "\n",
    "\n",
    "    #calculate fitness\n",
    "    #avgFit = 0\n",
    "    #for i in pop.keys():\n",
    "        #avgFit += (pop[i]/pop_size)*fitness[i]\n",
    "\n",
    "\n",
    "    #simulateSwitch(pop2, history ,generations, mutation_rate, pop_size, seq_length, fitness, fitnessB, alphabet)\n",
    "\n",
    "    plt.figure(num=None, figsize=(14, 14), dpi=80, facecolor='w', edgecolor='k')\n",
    "    plt.subplot2grid((3,2), (0,0), colspan=2)\n",
    "    stacked_trajectory_plot(history, generations, pop_size)\n",
    "    plt.subplot2grid((3,2), (1,0), colspan=2)\n",
    "    snp_trajectory_plot(history, seq_length, generations, pop_size)\n",
    "    plt.subplot2grid((3,2), (2,0))\n",
    "    diversity_plot(history, pop_size)\n",
    "    plt.subplot2grid((3,2), (2,1))\n",
    "    divergence_plot(history, base_haplotype, pop_size)\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#################################################################################################################################\n",
    "\"\"\"\n",
    "Simulate Wright-Fisher evolution for some amount of generations\n",
    "\"\"\"\n",
    "def simulate(pop, history ,generations, mutation_rate, pop_size, seq_length, fitness, alphabet):\n",
    "    clone_pop = dict(pop)\n",
    "    history.append(clone_pop)\n",
    "    for i in range(generations):\n",
    "        time_step(pop, mutation_rate, pop_size, seq_length, fitness, alphabet)\n",
    "        clone_pop = dict(pop)\n",
    "        history.append(clone_pop)\n",
    "\n",
    "def simulateSwitch(pop, history ,generations, mutation_rate, pop_size, seq_length, fitnessA, fitnessB, alphabet):\n",
    "    clone_pop = dict(pop)\n",
    "    history.append(clone_pop)\n",
    "\n",
    "    count = 0\n",
    "    for i in range(generations):\n",
    "        if (count < 1):\n",
    "            time_step(pop, mutation_rate, pop_size, seq_length, fitnessA, alphabet)\n",
    "            clone_pop = dict(pop)\n",
    "            history.append(clone_pop)\n",
    "            count += 1\n",
    "        else:\n",
    "            time_step(pop, mutation_rate, pop_size, seq_length, fitnessB, alphabet)\n",
    "            clone_pop = dict(pop)\n",
    "            history.append(clone_pop)\n",
    "            count += 1\n",
    "            if (count > 2):\n",
    "                count = 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Code to put into a separate class/function\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "Execuse one generation of the Wright-Fisher\n",
    "\"\"\"\n",
    "def time_step(pop, mutation_rate, pop_size, seq_length, fitness, alphabet):\n",
    "    mutation_step(pop, mutation_rate, pop_size, seq_length, alphabet)\n",
    "    offspring_step(pop, pop_size, fitness)\n",
    "\n",
    "\n",
    "#################################################################################################################################\n",
    "\"\"\"\n",
    "Below is the code responsible for mutation in the Wright-Fisher model, in order of function calls.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "First step of mutation -- get a count of the mutations that occur.\n",
    "\"\"\"\n",
    "def mutation_step(pop, mutation_rate, pop_size, seq_length, alphabet):\n",
    "    mutation_count = get_mutation_count(mutation_rate, pop_size, seq_length)\n",
    "    for i in range(mutation_count):\n",
    "        mutation_event(pop, pop_size, seq_length, alphabet)\n",
    "\n",
    "\"\"\"\n",
    "Draw mutation count from a poisson distribution with mean equal to average number of expected mutations.\n",
    "\"\"\"\n",
    "def get_mutation_count(mutation_rate, pop_size, seq_length):\n",
    "    mean = mutation_rate * pop_size * seq_length\n",
    "    return np.random.poisson(mean)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Function that find a random haplotype to mutate and adds that new mutant to the population. Reduces mutated population by 1.\n",
    "\"\"\"\n",
    "def mutation_event(pop, pop_size, seq_length, alphabet):\n",
    "    haplotype = get_random_haplotype(pop, pop_size)\n",
    "    if pop[haplotype] > 1:\n",
    "        pop[haplotype] -= 1\n",
    "        new_haplotype = get_mutant(haplotype, seq_length, alphabet)\n",
    "        if new_haplotype in pop:\n",
    "            pop[new_haplotype] += 1\n",
    "        else:\n",
    "            pop[new_haplotype] = 1\n",
    "\n",
    "\"\"\"\n",
    "Chooses a random haplotype in the population that will be returned.\n",
    "\"\"\"\n",
    "def get_random_haplotype(pop, pop_size):\n",
    "    haplotypes = list(pop.keys())\n",
    "    frequencies = [x/pop_size for x in pop.values()]\n",
    "    total = sum(frequencies)\n",
    "    frequencies = [x / total for x in frequencies]\n",
    "    return fast_choice(haplotypes, frequencies)\n",
    "    #return random.choices(haplotypes, weights=frequencies)[0]\n",
    "\n",
    "\"\"\"\n",
    "Receives the haplotype to be mutated and returns a new haplotype with a mutation with all neighbor mutations equally probable.\n",
    "\"\"\"\n",
    "def get_mutant(haplotype, seq_length, alphabet):\n",
    "    site = int(random.random()*seq_length)\n",
    "    possible_mutations = list(alphabet)\n",
    "    possible_mutations.remove(haplotype[site])\n",
    "    mutation = random.choice(possible_mutations)\n",
    "    new_haplotype = haplotype[:site] + mutation + haplotype[site+1:]\n",
    "    return new_haplotype\n",
    "\n",
    "\n",
    "#################################################################################################################################\n",
    "\"\"\"\n",
    "Below is the code responsible for offspring in the Wright-Fisher model, in order of function calls.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Gets the number of counts after an offspring step and stores them in the haplotype. If a population is reduced to zero then delete it.\n",
    "\"\"\"\n",
    "def offspring_step(pop, pop_size, fitness):\n",
    "    haplotypes = list(pop.keys())\n",
    "    counts = get_offspring_counts(pop, pop_size, fitness)\n",
    "    for (haplotype, count) in zip(haplotypes, counts):\n",
    "        if (count > 0):\n",
    "            pop[haplotype] = count\n",
    "        else:\n",
    "            del pop[haplotype]\n",
    "\n",
    "\"\"\"\n",
    "Returns the new population count for each haplotype given offspring counts weighted by fitness of haplotype\n",
    "\"\"\"\n",
    "def get_offspring_counts(pop, pop_size, fitness):\n",
    "    haplotypes = list(pop.keys())\n",
    "    frequencies = [pop[haplotype]/pop_size for haplotype in haplotypes]\n",
    "    fitnesses = [fitness[haplotype] for haplotype in haplotypes]\n",
    "    weights = [x * y for x,y in zip(frequencies, fitnesses)]\n",
    "    total = sum(weights)\n",
    "    weights = [x / total for x in weights]\n",
    "    return list(np.random.multinomial(pop_size, weights))\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "This is faster than numpy because numpy is dog water. Self-written code to choose an item from a list with some prob.\n",
    "\"\"\"\n",
    "def fast_choice(options, probs):\n",
    "    x = random.random()\n",
    "    cum = 0\n",
    "    for i, p in enumerate(probs):\n",
    "        cum += p\n",
    "        if x < cum:\n",
    "            return options[i]\n",
    "    return options[-1]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def get_distance(seq_a, seq_b):\n",
    "    diffs = 0\n",
    "    length = len(seq_a)\n",
    "    assert len(seq_a) == len(seq_b)\n",
    "    for chr_a, chr_b in zip(seq_a, seq_b):\n",
    "        if chr_a != chr_b:\n",
    "            diffs += 1\n",
    "    return diffs / float(length)\n",
    "\n",
    "def get_diversity(population, pop_size):\n",
    "    haplotypes = list(population.keys())\n",
    "    haplotype_count = len(haplotypes)\n",
    "    diversity = 0\n",
    "    for i in range(haplotype_count):\n",
    "        for j in range(haplotype_count):\n",
    "            haplotype_a = haplotypes[i]\n",
    "            haplotype_b = haplotypes[j]\n",
    "            frequency_a = population[haplotype_a] / float(pop_size)\n",
    "            frequency_b = population[haplotype_b] / float(pop_size)\n",
    "            frequency_pair = frequency_a * frequency_b\n",
    "            diversity += frequency_pair * get_distance(haplotype_a, haplotype_b)\n",
    "    return diversity\n",
    "\n",
    "def get_diversity_trajectory(history, pop_size):\n",
    "    trajectory = [get_diversity(generation, pop_size) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "def diversity_plot(history, pop_size):\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_diversity_trajectory(history, pop_size)\n",
    "    plt.plot(trajectory, \"#447CCD\")\n",
    "    plt.ylabel(\"diversity\")\n",
    "    plt.xlabel(\"generation\")\n",
    "\n",
    "def get_divergence(population, base_haplotype, pop_size):\n",
    "    haplotypes = population.keys()\n",
    "    divergence = 0\n",
    "    for haplotype in haplotypes:\n",
    "        frequency = population[haplotype] / float(pop_size)\n",
    "        divergence += frequency * get_distance(base_haplotype, haplotype)\n",
    "    return divergence\n",
    "\n",
    "def get_divergence_trajectory(history, base_haplotype, pop_size):\n",
    "    trajectory = [get_divergence(generation, base_haplotype, pop_size) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "def divergence_plot(history, base_haplotype, pop_size):\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_divergence_trajectory(history, base_haplotype, pop_size)\n",
    "    plt.plot(trajectory, \"#447CCD\")\n",
    "    plt.ylabel(\"divergence\")\n",
    "    plt.xlabel(\"generation\")\n",
    "\n",
    "def get_frequency(haplotype, generation, history, pop_size):\n",
    "    pop_at_generation = history[generation]\n",
    "    if haplotype in pop_at_generation:\n",
    "        return pop_at_generation[haplotype]/float(pop_size)\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "def get_trajectory(haplotype, generations, history, pop_size):\n",
    "    trajectory = [get_frequency(haplotype, gen, history, pop_size) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_haplotypes(history):\n",
    "    haplotypes = set()\n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            haplotypes.add(haplotype)\n",
    "    return haplotypes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stacked_trajectory_plot(history, generations, pop_size, xlabel=\"generation\"):\n",
    "    colors_lighter = [\"#A567AF\", \"#8F69C1\", \"#8474D1\", \"#7F85DB\", \"#7F97DF\", \"#82A8DD\", \"#88B5D5\", \"#8FC0C9\", \"#97C8BC\", \"#A1CDAD\", \"#ACD1A0\", \"#B9D395\", \"#C6D38C\", \"#D3D285\", \"#DECE81\", \"#E8C77D\", \"#EDBB7A\", \"#EEAB77\", \"#ED9773\", \"#EA816F\", \"#E76B6B\"]\n",
    "    mpl.rcParams['font.size']=18\n",
    "    haplotypes = get_all_haplotypes(history)\n",
    "    trajectories = [get_trajectory(haplotype, generations, history, pop_size) for haplotype in haplotypes]\n",
    "    plt.stackplot(range(generations), trajectories, colors=colors_lighter)\n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)\n",
    "\n",
    "def get_snp_frequency(site, generation, history, pop_size):\n",
    "    minor_allele_frequency = 0.0\n",
    "    pop_at_generation = history[generation]\n",
    "    for haplotype in pop_at_generation.keys():\n",
    "        allele = haplotype[site]\n",
    "        frequency = pop_at_generation[haplotype] / float(pop_size)\n",
    "        if allele != \"0\":\n",
    "            minor_allele_frequency += frequency\n",
    "    return minor_allele_frequency\n",
    "\n",
    "def get_snp_trajectory(site, generations, history, pop_size):\n",
    "    trajectory = [get_snp_frequency(site, gen, history, pop_size) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_snps(history, seq_length):\n",
    "    snps = set()\n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            for site in range(seq_length):\n",
    "                if haplotype[site] != \"0\":\n",
    "                    snps.add(site)\n",
    "    return snps\n",
    "\n",
    "def snp_trajectory_plot(history, seq_length, generations, pop_size, xlabel=\"generation\"):\n",
    "    colors = [\"#781C86\", \"#571EA2\", \"#462EB9\", \"#3F47C9\", \"#3F63CF\", \"#447CCD\", \"#4C90C0\", \"#56A0AE\", \"#63AC9A\", \"#72B485\", \"#83BA70\", \"#96BD60\", \"#AABD52\", \"#BDBB48\", \"#CEB541\", \"#DCAB3C\", \"#E49938\", \"#E68133\", \"#E4632E\", \"#DF4327\", \"#DB2122\"]\n",
    "    mpl.rcParams['font.size']=18\n",
    "    snps = get_all_snps(history, seq_length)\n",
    "    trajectories = [get_snp_trajectory(snp, generations, history, pop_size) for snp in snps]\n",
    "    data = []\n",
    "    for trajectory, color in zip(trajectories, itertools.cycle(colors)):\n",
    "        data.append(range(generations))\n",
    "        data.append(trajectory)\n",
    "        data.append(color)\n",
    "    fig = plt.plot(*data)\n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = mainfunc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y = mainfunc()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
